java/java-impl/src/com/intellij/codeInsight/FunctionalInterfaceSuggester.java

```diff
@@ -260,22 +253,28 @@ else if (expression instanceof PsiMethodReferenceExpression) {
 
                 for (int i = 0; i < targetMethodParameters.length; i++) {
                   left[i + offset] = parameters[i + offset].getType();
-                  right[i + offset] = partialSubstitutor.substitute(targetMethodParameters[i].getType());
+                  right[i + offset] = targetMethodParameters[i].getType();
                 }
 
+                final PsiSubstitutor partialSubstitutor = getPartialSubstitutor(referenceExpression, method);
+
                 left[parameters.length] = method.isConstructor() ? qualifierType : partialSubstitutor.substitute(method.getReturnType());
```

Comment:
partial substitution here is applied to the left type and when you pass partial substitutor to helper it would be applied to the `returnType` as well. Though I don't have a sample where it would lead to problems, I'd rather have previous behaviour as I don't see why it should be changed at the first place

Reply:
Thank you for the review.
Are you talking about [this code](https://github.com/JetBrains/intellij-community/blob/master/java/java-psi-impl/src/com/intellij/psi/impl/source/resolve/graphInference/InferenceSession.java#L90)?

Reply:
yes

Reply:
Ok, I don't fully understand it, so I will trust your judgement. I will bring back the substitution of `right` back here and remove the `partialSubstitutor` from the argument list of `PsiResolveHelper#inferTypeArguments`

Reply:
I removed `partialSubstitutor` from the argument list